File: frontend/src/lib/stores/plan.ts
import { writable, derived, get, type Readable } from 'svelte/store';
import { fetchWithTimeout } from '$lib/utils';

export interface PlanFeatures {
	basic_analytics: boolean;
	analytics: boolean;
	badge_forced: boolean;
	courses: boolean;
	api: boolean;
	custom_domain: boolean;
	newsletter: boolean;
	discussions: boolean;
	pricing: boolean;
}

export interface PlanConfig {
	managed: boolean;
	plan: string;
	features: PlanFeatures;
	storage_limit_mb?: number;
	custom_domain?: string;
	demo_mode?: boolean;
}

const defaultConfig: PlanConfig = {
	managed: false,
	plan: 'self-hosted',
	features: {
		basic_analytics: true,
		analytics: true,
		badge_forced: false,
		// Courses default OFF self-hosted: public /courses/[slug] route is not implemented
		// in this repo (cloud only). Backend authority via /api/plan can flip this on
		// when FACET_FEATURE_COURSES=true is set on the server.
		courses: false,
		api: true,
		custom_domain: true,
		newsletter: true,
		discussions: true,
		pricing: true
	}
};

export const planConfig = writable<PlanConfig>(defaultConfig);
export const planInitialized = writable<boolean>(false);

export const isManaged = derived(planConfig, ($config) => $config.managed);
export const currentPlan = derived(planConfig, ($config) => $config.plan);
export const planFeatures = derived(planConfig, ($config) => $config.features);
export const brandName = derived(planConfig, ($config) => ($config.managed ? 'Facet Cloud' : 'Facet'));
export const isDemoMode = derived(planConfig, ($config) => $config.demo_mode === true);

// Features that are not fully implemented in the self-hosted build and must
// honor the explicit flag value even when not in managed mode. Adding a feature
// here means "self-hosted does not unconditionally get this feature".
const selfHostedOptInFeatures: ReadonlySet<keyof PlanFeatures> = new Set(['courses']);

// Non-reactive version for use in script blocks
export function checkFeature(feature: keyof PlanFeatures): boolean {
	const config = get(planConfig);
	if (selfHostedOptInFeatures.has(feature)) return config.features[feature] === true;
	if (!config.managed) return true;
	return config.features[feature] ?? false;
}

// Reactive version for use in templates
export function hasFeature(feature: keyof PlanFeatures): Readable<boolean> {
	return derived(planConfig, ($config) => {
		if (selfHostedOptInFeatures.has(feature)) return $config.features[feature] === true;
		if (!$config.managed) return true;
		return $config.features[feature] ?? false;
	});
}

export async function initPlan() {
	try {
		const response = await fetchWithTimeout('/api/plan');
		if (response.ok) {
			const data: PlanConfig = await response.json();
			planConfig.set(data);
		}
	} catch {
		// On error, keep defaults (self-hosted with all features)
	}
	planInitialized.set(true);
}

/**
 * Initialize plan config from SSR data (if available)
 * This eliminates FOUC by setting the correct plan state before first render
 */
export function initPlanFromSSR(data: PlanConfig | null) {
	if (data) {
		planConfig.set(data);
		planInitialized.set(true);
	}
}

File: backend/services/plan.go
package services

import (
	"os"
	"strconv"
)

// PlanTier represents the subscription plan tier
type PlanTier string

const (
	PlanSelfHosted PlanTier = "self-hosted"
)

// PlanFeatures represents which features are available for the current plan
type PlanFeatures struct {
	BasicAnalytics bool `json:"basic_analytics"`
	Analytics      bool `json:"analytics"`
	BadgeForced    bool `json:"badge_forced"`
	Courses        bool `json:"courses"`
	API            bool `json:"api"`
	CustomDomain   bool `json:"custom_domain"`
	Newsletter     bool `json:"newsletter"`
	Discussions    bool `json:"discussions"`
	Pricing        bool `json:"pricing"`
}

// PlanConfig holds the managed mode configuration derived from environment variables.
// Self-hosted instances get all features unlocked by default.
type PlanConfig struct {
	Managed  bool         `json:"managed"`
	Plan     PlanTier     `json:"plan"`
	Features PlanFeatures `json:"features"`
}

// LoadPlanConfig reads plan configuration from environment variables.
// When FACET_MANAGED is not set or false, returns a self-hosted config with all features unlocked.
// Self-hosted is the expected mode for this upstream repo; managed support is minimal for compatibility.
//
// Courses is OFF by default self-hosted because the public /courses/[slug] route is
// not implemented in this repo (cloud has it). Operators who want admin-only course
// authoring (e.g., headless integrations) can opt in with FACET_FEATURE_COURSES=true.
func LoadPlanConfig() *PlanConfig {
	managed := envBool("FACET_MANAGED", false)

	if !managed {
		return &PlanConfig{
			Managed: false,
			Plan:    PlanSelfHosted,
			Features: PlanFeatures{
				BasicAnalytics: true,
				Analytics:      true,
				BadgeForced:    false,
				Courses:        envBool("FACET_FEATURE_COURSES", false),
				API:            true,
				CustomDomain:   true,
				Newsletter:     true,
				Discussions:    true,
				Pricing:        true,
			},
		}
	}

	// Managed mode: minimal support for compatibility.
	// In practice, self-hosted instances never set FACET_MANAGED=true.
	return &PlanConfig{
		Managed: true,
		Plan:    PlanSelfHosted,
		Features: PlanFeatures{
			BasicAnalytics: true,
			Analytics:      true,
			BadgeForced:    false,
			Courses:        true,
			API:            true,
			Newsletter:     true,
			Discussions:    true,
			Pricing:        true,
		},
	}
}

// IsManaged returns whether this instance is running in managed (cloud) mode
func (c *PlanConfig) IsManaged() bool {
	return c.Managed
}

// HasFeature checks if a specific feature is available.
// Self-hosted instances always return true for all features.
func (c *PlanConfig) HasFeature(feature string) bool {
	switch feature {
	case "basic_analytics":
		return c.Features.BasicAnalytics
	case "analytics":
		return c.Features.Analytics
	case "courses":
		return c.Features.Courses
	case "api":
		return c.Features.API
	case "custom_domain":
		return c.Features.CustomDomain
	case "newsletter":
		return c.Features.Newsletter
	case "discussions":
		return c.Features.Discussions
	case "pricing":
		return c.Features.Pricing
	default:
		// Self-hosted gets everything; managed defaults to false for unknown features
		return !c.Managed
	}
}

// SignupAllowed returns whether public signup/registration is allowed.
// Self-hosted instances always allow signup.
func (c *PlanConfig) SignupAllowed() bool {
	return !c.Managed
}

func envBool(key string, fallback bool) bool {
	v := os.Getenv(key)
	if v == "" {
		return fallback
	}
	b, err := strconv.ParseBool(v)
	if err != nil {
		return fallback
	}
	return b
}
